File: Wire-iOS/Sources/UserInterface/LaunchImage/LaunchImageViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "LaunchImageViewController.h"

@import PureLayout;
#import "Constants.h"
@import WireExtensionComponents;
#import "Settings.h"
#import "Wire-Swift.h"

@interface LaunchImageViewController ()

@property (nonatomic) UIView *contentView;
@property (nonatomic) BOOL shouldShowLoadingScreenOnViewDidLoad;
@property (nonatomic) UILabel *loadingScreenLabel;
@property (nonatomic) ProgressSpinner *activityIndicator;

@end

@implementation LaunchImageViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    NSArray *loadedObjects = [[UINib nibWithNibName:@"LaunchScreen" bundle:nil] instantiateWithOwner:nil options:nil];

    UIView *nibView = loadedObjects.firstObject;
    nibView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:nibView];
    self.contentView = nibView;

    self.activityIndicator = [[ProgressSpinner alloc] init];
    self.activityIndicator.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.activityIndicator];
    
    self.loadingScreenLabel = [[UILabel alloc] initForAutoLayout];
    self.loadingScreenLabel.font = [UIFont systemFontOfSize:12];
    self.loadingScreenLabel.textColor = [UIColor whiteColor];
    
    self.loadingScreenLabel.text = [NSLocalizedString(@"migration.please_wait_message", @"") uppercaseStringWithLocale:[NSLocale currentLocale]];
    self.loadingScreenLabel.hidden = YES;

    [self.view addSubview:self.loadingScreenLabel];
    
    // Constraints
    [self.contentView autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
    
    [self.loadingScreenLabel autoAlignAxisToSuperviewMarginAxis:ALAxisVertical];
    [self.loadingScreenLabel autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:48];
    
    [self.activityIndicator autoAlignAxisToSuperviewMarginAxis:ALAxisVertical];
    [self.activityIndicator autoPinEdge:ALEdgeBottom toEdge:ALEdgeTop ofView:self.loadingScreenLabel withOffset:-24];
    
    // Start the spinner in case of it was requested right after the init
    if (self.shouldShowLoadingScreenOnViewDidLoad) {
        [self showLoadingScreen];
    }
}

- (BOOL)prefersStatusBarHidden
{
    return YES;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return self.wr_supportedInterfaceOrientations;
}

- (void)showLoadingScreen
{
    self.shouldShowLoadingScreenOnViewDidLoad = YES;
    
    self.loadingScreenLabel.hidden = NO;
    [self.activityIndicator startAnimation:nil];
}

- (void)hideLoadingScreen
{
    [self.activityIndicator stopAnimation:nil];
    self.loadingScreenLabel.hidden = YES;
}

@end

File: Wire-iOS/Sources/UserInterface/LaunchImage/LaunchImageViewController.swift
//
// Wire
// Copyright (C) 2019 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

/// Replicates the launch screen to prevent the black screen being visible, cause of later UI initialization
class LaunchImageViewController: UIViewController {

    private var shouldShowLoadingScreenOnViewDidLoad = false

    private var contentView: UIView!
    private let loadingScreenLabel = UILabel()
    private let activityIndicator = ProgressSpinner()

    /// Convenience method for showing the @c activityIndicator and @c loadingScreenLabel and start the spinning animation
    func showLoadingScreen() {
        shouldShowLoadingScreenOnViewDidLoad = true
        loadingScreenLabel.isHidden = false
        activityIndicator.startAnimation()
    }

    /// Convenience method for hiding all the animation related functionality
    func hideLoadingScreen() {
        activityIndicator.stopAnimation()
        loadingScreenLabel.isHidden = true
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let loadedObjects = UINib(nibName: "LaunchScreen", bundle: nil).instantiate(withOwner: nil, options: nil)

        let nibView = loadedObjects.first as? UIView
        nibView?.translatesAutoresizingMaskIntoConstraints = false
        if let nibView = nibView {
            view.addSubview(nibView)
        }
        if let nibView = nibView {
            contentView = nibView
        }

        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(activityIndicator)

        loadingScreenLabel.font = .systemFont(ofSize: 12)
        loadingScreenLabel.textColor = .white

        loadingScreenLabel.text = "migration.please_wait_message".localized.uppercased(with: NSLocale.current)
        loadingScreenLabel.isHidden = true

        view.addSubview(loadingScreenLabel)

        createConstraints()

        // Start the spinner in case of it was requested right after the init
        if shouldShowLoadingScreenOnViewDidLoad {
            showLoadingScreen()
        }
    }

    private func createConstraints() {
        [contentView, loadingScreenLabel, activityIndicator].prepareForLayout()

        var constraints: [NSLayoutConstraint] = []

        constraints += [contentView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
                        contentView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
                        contentView.topAnchor.constraint(equalTo: view.topAnchor),
                        contentView.bottomAnchor.constraint(equalTo: view.bottomAnchor)]

        constraints.append(loadingScreenLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor))
        constraints.append(loadingScreenLabel.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -40))

        constraints.append(activityIndicator.centerXAnchor.constraint(equalTo: view.centerXAnchor))
        constraints.append(activityIndicator.bottomAnchor.constraint(equalTo: loadingScreenLabel.topAnchor, constant: -24))

        NSLayoutConstraint.activate(constraints)
    }

    override var prefersStatusBarHidden: Bool {
        return true
    }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return wr_supportedInterfaceOrientations
    }
}
